test/extended/prometheus/prometheus.go

```diff
@@ -278,6 +279,150 @@ var _ = g.Describe("[sig-instrumentation][Late] OpenShift alerting rules", func(
+			"KubeContainerWaiting",
+
+			// some metrics may be absent.
+			"KubeDaemonSetRolloutStuck",
+
+			// metrics may be absent when there are no HPA.
+			"KubeHpaMaxedOut",
+			"KubeHpaReplicasMismatch",
+
+			// kube_job_failed may be absent.
+			"KubeJobFailed",
+
+			// kube_node_spec_taint selector may return empty results.
+			"KubeNodeUnreachable",
+
+			// kube_persistentvolumeclaim_access_mode selector may return empty results.
+			"KubePersistentVolumeFillingUp",
+
+			// kube_pod_container_status_waiting_reason selector may return empty results.
+			"KubePodCrashLooping",
+
+			// kube_state_metrics_list_total selector may return empty results.
+			"KubeStateMetricsListErrors",
+			"KubeStateMetricsWatchErrors",
+
+			// mapi_machine_created_timestamp_seconds selector may return empty results.
+			"MachineNotYetDeleted",
+			"MachineWithNoRunningPhase",
+
+			// kube_pod_container_status_last_terminated_reason selector may return empty results.
+			"MultipleContainersOOMKilled",
+
+			// metrics may be absent if RAID is disabled.
+			"NodeRAIDDegraded",
+			"NodeRAIDDiskFailure",
+
+			// metrics may be absent if remote write isn't used.
+			"PrometheusRemoteStorageFailures",
+			"PrometheusRemoteWriteBehind",
+			"PrometheusRemoteWriteDesiredShards",
+
+			// openshift_samples_failed_imagestream_import_info may be absent.
+			"SamplesImagestreamImportFailing",
+			"SamplesRetriesMissingOnImagestreamImportFailing",
+
+			// cluster_operator_conditions selector may return empty results.
+			"SimpleContentAccessNotAvailable",
+
+			// metrics may be absent is there are no tech preview features enabled.
+			"TechPreviewNoUpgrade",
+
+			// thanos query error metrics may be absent.
+			"ThanosQueryGrpcClientErrorRate",
+			"ThanosQueryHttpRequestQueryErrorRateHigh",
+			"ThanosQueryHttpRequestQueryRangeErrorRateHigh",
+			"ThanosSidecarBucketOperationsFailed",
+
+			// cluster_version_available_updates may be empty.
+			"UpdateAvailable",
+		)
+
+		promClient := oc.NewPrometheusClient(context.TODO())
+
+		err := helper.ForEachAlertingRule(alertingRules, func(alert promv1.AlertingRule) sets.String {
+			if exceptions.Has(alert.Name) {
+				return nil
+			}
+
+			// Remove backslashs added by the HTTP API since they can't be parsed by
+			// the promQL parser.
+			query := strings.Replace(alert.Query, "\\", "", -1)
+			expr, err := parser.ParseExpr(query)
+			if err != nil {
+				return sets.NewString("couldn't parse query")
+			}
+
+			var exprs []string
+			parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+				switch n := node.(type) {
+				case *parser.VectorSelector:
+					exprs = append(exprs, fmt.Sprintf("absent(%s)", n))
+				}
+				return nil
+			})
+			q := strings.Join(exprs, " or ")
+
+			resp, err := helper.RunQuery(context.TODO(), promClient, q)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			if len(resp.Data.Result) > 0 {
+				return sets.NewString("one of the metrics used in the query is absent")
```

Comment:
would we be able to point out the metric? and print the query?

Reply:
doing that might be a bit too expensive. Currently, in order to avoid sending too many requests to Prometheus, I am transforming the query into:
```
absent(metrics1) or absent(metrics2) or absent(metrics3) or ...
```
IIRC we are currently checking 300+ alerts and if we were to check each metric used by an alert one by one that would amount to a lot of requests